controlpanel/api/models/access_to_s3bucket.py

```diff
@@ -45,27 +47,35 @@ class Meta:
     def has_readwrite_access(self):
         return self.access_level == self.READWRITE
 
-    def aws_name(self):
+    @property
+    def iam_role_name(self):
         raise NotImplementedError
 
-    def save(self, *args, **kwargs):
-        super().save(*args, **kwargs)
+    def grant_bucket_access(self):
+        raise NotImplementedError
 
-        with self.policy_class.load(self.aws_name()) as policy:
-            policy.revoke_access(self.s3bucket.arn)
-            for resource in self.resources:
-                policy.grant_access(resource, self.access_level)
+    def revoke_bucket_access(self):
+        raise NotImplementedError
 
+    def save(self, *args, **kwargs):
+        super().save(*args, **kwargs)
+        self.grant_bucket_access()
         return self
 
     def delete(self, *args, **kwargs):
-        with self.policy_class.load(self.aws_name()) as policy:
-            for resource in self.resources:
-                policy.revoke_access(resource)
-
```

Comment:
As discussed offline earlier (and for the record) the bit on the right makes more sense to me, but at the same time it would be also useful to understand what the original logic was doing and if we're missing something important here.